Mathlib/GroupTheory/Subgroup/Pointwise.lean

```diff
@@ -74,24 +74,44 @@ theorem closure_toSubmonoid (S : Set G) :
 #align add_subgroup.closure_to_add_submonoid AddSubgroup.closure_toAddSubmonoid
 
 /-- For subgroups generated by a single element, see the simpler `zpow_induction_left`. -/
-@[to_additive "For additive subgroups generated by a single element, see the simpler
-`zsmul_induction_left`."]
-theorem closure_induction_left {p : G → Prop} {x : G} (h : x ∈ closure s) (H1 : p 1)
-    (Hmul : ∀ x ∈ s, ∀ (y), p y → p (x * y)) (Hinv : ∀ x ∈ s, ∀ (y), p y → p (x⁻¹ * y)) : p x :=
-  let key := (closure_toSubmonoid s).le
-  Submonoid.closure_induction_left (key h) H1 fun x hx => hx.elim (Hmul x) fun hx y hy =>
-    inv_inv x ▸ Hinv x⁻¹ hx y hy
+@[to_additive (attr := elab_as_elim)
+  "For additive subgroups generated by a single element, see the simpler
+  `zsmul_induction_left`."]
+theorem closure_induction_left {p : (x : G) → x ∈ closure s → Prop} (H1 : p 1 (one_mem _))
+    (mul_left : ∀ x (hx : x ∈ s), ∀ (y) hy, p y hy → p (x * y) (mul_mem (subset_closure hx) hy))
+    (mul_left_inv : ∀ x (hx : x ∈ s), ∀ (y) hy, p y hy →
+      p (x⁻¹ * y) (mul_mem (inv_mem (subset_closure hx)) hy))
+    {x : G} (h : x ∈ closure s) : p x h := by
+  revert h
+  simp_rw [← mem_toSubmonoid, closure_toSubmonoid] at *
+  intro h
+  induction h using Submonoid.closure_induction_left with
+  | one => exact H1
+  | mul_left x hx y hy ih =>
+    cases hx with
+    | inl hx => exact mul_left _ hx _ hy ih
+    | inr hx => simpa only [inv_inv] using mul_left_inv _ hx _ hy ih
 #align subgroup.closure_induction_left Subgroup.closure_induction_left
 #align add_subgroup.closure_induction_left AddSubgroup.closure_induction_left
 
 /-- For subgroups generated by a single element, see the simpler `zpow_induction_right`. -/
-@[to_additive "For additive subgroups generated by a single element, see the simpler
-`zsmul_induction_right`."]
-theorem closure_induction_right {p : G → Prop} {x : G} (h : x ∈ closure s) (H1 : p 1)
-    (Hmul : ∀ (x), ∀ y ∈ s, p x → p (x * y)) (Hinv : ∀ (x), ∀ y ∈ s, p x → p (x * y⁻¹)) : p x :=
-  let key := (closure_toSubmonoid s).le
-  Submonoid.closure_induction_right (key h) H1 fun x y hy => hy.elim (Hmul x y) fun hy hx =>
-    inv_inv y ▸ Hinv x y⁻¹ hy hx
+@[to_additive (attr := elab_as_elim)
+  "For additive subgroups generated by a single element, see the simpler
+  `zsmul_induction_right`."]
+theorem closure_induction_right {p : (x : G) → x ∈ closure s → Prop} (H1 : p 1 (one_mem _))
+    (mul_right : ∀ (x) hx, ∀ y (hy : y ∈ s), p x hx → p (x * y) (mul_mem hx (subset_closure hy)))
+    (mul_right_inv : ∀ (x) hx, ∀ y (hy : y ∈ s), p x hx →
+      p (x * y⁻¹) (mul_mem hx (inv_mem (subset_closure hy))))
+    {x : G} (h : x ∈ closure s) : p x h := by
```

Comment:
Can't you also prove this one using the opposite product and recycle the `closure_induction_left` proof?

Reply:
Sure, done

Reply:
Very cool, thanks!